google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java

```diff
@@ -209,6 +240,10 @@ public String getTopicNameString() {
     messagesWaiter.incrementPendingMessages(1);
 
     if (!batchesToSend.isEmpty()) {
+      // TODO: if this is not an ordering keys scenario, will this do anything?
+      publishAllWithoutInflight();
+
+      // TODO: if this is an ordering keys scenario, is this safe without messagesBatchLock?
       for (final OutstandingBatch batch : batchesToSend) {
         logger.log(Level.FINER, "Scheduling a batch for immediate sending.");
         executor.execute(
```

Comment:
It seems that this execute() call was removed in the original implementation of ordering keys, but was added back when master is merged to this branch. Please see https://github.com/googleapis/google-cloud-java/commit/3bdbf7701bdb29a3a70a667d2a624d7fae0e5442#diff-c482c19af8c46aea82546fa17236dfd2

publishOutstandingBatch() should not be called by the executor here because it can reorder sequence of invoking the function. Inside publishOutstandingBatch(), we run an executor, a regular one for non-ordered messages and the sequential executor for ordered messages.

Reply:
Woops.  I'll change it back.  What do you think about moving the `publish*` methods inside the lock?